README
Currency-Exchange-IMF

The module was created to extract in a easy way IMF SDR exchange rates by using IMF query tool.


INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build installdeps
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Currency::Exchange::IMF

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        https://rt.cpan.org/NoAuth/Bugs.html?Dist=Currency-Exchange-IMF

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Currency-Exchange-IMF

    CPAN Ratings
        https://cpanratings.perl.org/d/Currency-Exchange-IMF

    Search CPAN
        https://metacpan.org/release/Currency-Exchange-IMF


LICENSE AND COPYRIGHT

This software is Copyright (c) 2020 by Guillermo Martinez.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

USAGE

Usage: currency-exchange-imf -c=USD,EUR -f=2020-01-30
Usage: currency-exchange-imf -c=USD,EUR -f=2020-01-30 -t=2020-02-05

Perl Module Usage:
    my $imf = Currency::Exchange::IMF->new('targetCurrencies' => ,\@currencies, 'sourceCurrency' => 'SDR', 'fromDate' => $from, 'toDate' => $to);

    my $data = $imf->getRates;
Options:
    -c|--currency: comma seperated list of currencies codes Ex: usd,eur
    -f|--from:     Start date from which the rates will be fetched
    -t|--to (optional): End date to stop fetching rates.

Supported currencies:
    CNY EUR JPY GBP USD DZD AUD ATS BHD BEF VEF VES BWP BRL BND CAD CLP COP CYP CZK DKK DEM EEK FIM FRF GRD HUF
    ISK INR IDR IRR IEP ILS ITL KZT KRW KWD LYD LUF MYR MTL MUR MXN NPR NLG NZD NOK OMR PKR PEN PHP PLN PTE QAR
    RUB SAR SGD SKK SIT ZAR ESP LKR SEK CHF THB TTD TND AED UYU VEB

Output:
    DATE,SOURCE_CURRENCY,TARGET_CURRENCY,EXCHANGE_RATE
    04-Feb-2020,SDR,USD,1.37635
    04-Feb-2020,SDR,EUR,1.24579
    05-Feb-2020,SDR,USD,1.37542
    05-Feb-2020,SDR,EUR,1.24777
    06-Feb-2020,SDR,USD,1.37356
    06-Feb-2020,SDR,EUR,1.24835
    03-Feb-2020,SDR,USD,1.37711
    03-Feb-2020,SDR,EUR,1.24446


Changes
Revision history for Currency-Exchange-IMF

0.01    Date/time
        First version, released on an unsuspecting world.



Build.PL
use strict;
use warnings;
use Module::Build;

use 5.008;

my $builder = Module::Build->new(
    module_name         => 'Currency::Exchange::IMF',
    license             => 'perl',
    dist_author         => 'Guillermo Martinez <guillermo.marcial@gmail.com>',
    dist_abstract       => 'IMF SDR Exchange rate extraction',
    create_makefile_pl  => 'traditional',
    script_files        => 'script/currency-exchange-imf',
    create_readme       => 0,
    requires            => {
        'Modern::Perl'                => '0',
        'Moose'                       => '0',
        'DateTime::Format::DateParse' => '0',
        'WWW::Mechanize'              => '0',
        'File::Util::Tempdir'         => '0',
        'File::Util'                  => '0',
        'String::Random'              => '0',
        'XML::LibXML::Simple'         => '0',
        'Try::Tiny'                   => '0',
        'LWP::Protocol::https'        => '0'
    },
    build_requires      => {
        'Test::More'          => '0.47',
    },
);

$builder->create_build_script();


Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;


my %WriteMakefileArgs = (
    NAME             => 'Currency::Exchange::IMF',
    AUTHOR           => q{Guillermo Martinez <guillermo.marcial@gmail.com>},
    VERSION_FROM     => 'lib/Currency/Exchange/IMF.pm',
    ABSTRACT_FROM    => 'lib/Currency/Exchange/IMF.pm',
    LICENSE          => 'artistic_2',
    MIN_PERL_VERSION => '5.006',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => '0',
    },
    TEST_REQUIRES => {
        'Test::More' => '0',
    },
    PREREQ_PM => {
        'Modern::Perl'                => '0',
        'Moose'                       => '0',
        'DateTime::Format::DateParse' => '0',
        'WWW::Mechanize'              => '0',
        'File::Util::Tempdir'         => '0',
        'File::Util'                  => '0',
        'String::Random'              => '0',
        'XML::LibXML::Simple'         => '0',
        'Try::Tiny'                   => '0',
        'LWP::Protocol::https'        => '0'
    },
    EXE_FILES => [ qw(script/currency-exchange-imf) ],
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Currency-Exchange-IMF-*' },
);

# Compatibility with old versions of ExtUtils::MakeMaker
unless (eval { ExtUtils::MakeMaker->VERSION('6.64'); 1 }) {
    my $test_requires = delete $WriteMakefileArgs{TEST_REQUIRES} || {};
    @{$WriteMakefileArgs{PREREQ_PM}}{keys %$test_requires} = values %$test_requires;
}

unless (eval { ExtUtils::MakeMaker->VERSION('6.55_03'); 1 }) {
    my $build_requires = delete $WriteMakefileArgs{BUILD_REQUIRES} || {};
    @{$WriteMakefileArgs{PREREQ_PM}}{keys %$build_requires} = values %$build_requires;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
    unless eval { ExtUtils::MakeMaker->VERSION('6.52'); 1 };
delete $WriteMakefileArgs{MIN_PERL_VERSION}
    unless eval { ExtUtils::MakeMaker->VERSION('6.48'); 1 };
delete $WriteMakefileArgs{LICENSE}
    unless eval { ExtUtils::MakeMaker->VERSION('6.31'); 1 };

WriteMakefile(%WriteMakefileArgs);


lib/Currency/Exchange/IMF.pm
package Currency::Exchange::IMF;

use 5.006;
use strict;
use warnings;
use Moose;
use DateTime::Format::DateParse;
use WWW::Mechanize;
use File::Util::Tempdir qw(get_tempdir get_user_tempdir);
use File::Util;
use String::Random;
use XML::LibXML::Simple   qw(XMLin);

=head1 NAME

Currency::Exchange::IMF - Module to extract SDR exchange rates from IMF website.
The fetch process is slow, the module is intended to be used as an extraction tool to store the rates into a database
and nothing more.

=head1 VERSION

Version 0.01

=cut

our $VERSION = '1.00';

##  Attributes ##
has 'targetCurrencies' => (
    'is' => 'rw',
    'isa'     => 'ArrayRef',
    'default' => sub { ['USD'] });
has 'sourceCurrency' => ('is' => 'rw', default => 'SDR');
has 'fromDate'  => ('is' => 'rw');
has 'toDate'  => ('is' => 'rw');
has 'url' => ('is' => 'rw', default => 'https://www.imf.org/external/np/fin/ert/GUI/Pages/');

my $tmpDir = get_tempdir();
my $randomString = String::Random->new;
=head1 SYNOPSIS

The module uses IMF Exchange Query Tool to fetch SDR exchange rates.

How to use:

    my $imf = Currency::Exchange::IMF->new('targetCurrencies' => ['USD', 'EUR'], 'sourceCurrency' => 'SDR', 'fromDate' => '2020-02-03', 'toDate' => '2020-02-03');
    my %rates = $imf->getRates;
    ...

This would return somthing like this:
    {
        03-Feb-2020   {
            EUR   1.24446,
            USD   1.37711
        }
    }

=head1 EXPORT

A list of functions that can be exported.

=head1 SUBROUTINES/METHODS

=head2 getRates

    Used to fetch rates

=cut
sub getRates{
    my $self = shift;

    my $tmpFile = $tmpDir.File::Util->SL.$randomString->randpattern("cccccccccccccccccccc").'.xml';
    my $browser = WWW::Mechanize->new();

    $browser->get($self->buildUrl);
    $browser->follow_link(url=>'ReportData.aspx?Type=XML');
    $browser->save_content($tmpFile);

    my $parser = XML::LibXML->new();
    my $data = XMLin($tmpFile);
    unlink($tmpFile); #Remove temporal file
    return generateHash($data);
}

sub generateHash {
    my ($data) = @_;
    my %structure;

    #Generate Body
    if(ref($data->{'EFFECTIVE_DATE'}) eq 'ARRAY'){
        foreach my $efectiveRate (@{$data->{'EFFECTIVE_DATE'}}){
            $structure{$efectiveRate->{'VALUE'}}={};
            if(ref($efectiveRate->{'RATE_VALUE'}) eq 'ARRAY') {
                foreach (@{$efectiveRate->{'RATE_VALUE'}}) {
                    $structure{$efectiveRate->{'VALUE'}}{$_->{'ISO_CHAR_CODE'}} = $_->{'content'};
                }
            }else{
                $structure{$efectiveRate->{'VALUE'}}{$efectiveRate->{'RATE_VALUE'}->{'ISO_CHAR_CODE'}} = $efectiveRate->{'RATE_VALUE'}->{'content'};
            }
        }
    }else{
        $structure{$data->{'EFFECTIVE_DATE'}->{'VALUE'}}={};
        if(ref($data->{'EFFECTIVE_DATE'}->{'RATE_VALUE'}) eq 'ARRAY'){
            foreach(@{$data->{'EFFECTIVE_DATE'}->{'RATE_VALUE'}}){
                $structure{$data->{'EFFECTIVE_DATE'}->{'VALUE'}}{$_->{'ISO_CHAR_CODE'}}=$_->{'content'};
            }
        }else{
            $structure{$data->{'EFFECTIVE_DATE'}->{'VALUE'}}{$data->{'EFFECTIVE_DATE'}->{'RATE_VALUE'}->{'ISO_CHAR_CODE'}}=$data->{'EFFECTIVE_DATE'}->{'RATE_VALUE'}->{'content'};
        }
    }

    return \%structure;
}

sub buildUrl{
    my $self = shift;
    return $self->url.'Report.aspx?'.$self->getQueryParams();
}

sub getQueryParams{
    my $self = shift;
    my $from = undef;
    my $to = undef;

    #Set time to IMF Format
    $from = DateTime::Format::DateParse->parse_datetime($self->fromDate, 'UTC');
    $to = DateTime::Format::DateParse->parse_datetime($self->toDate, 'UTC');

    my $yearOne = DateTime->new(
        year       => 1,
        month      => 1,
        day        => 1,
        hour       => 0,
        minute     => 0,
        second     => 0,
        nanosecond => 0,
        time_zone  => 'UTC',
    );
    my $fromSecondsSinceYearOne = $yearOne->delta_days($from)->delta_days() * 86400;
    my $toSecondsSinceYearOne = $yearOne->delta_days($to)->delta_days() * 86400;
    my $target = 'CU='.$self->serealizeCurrency;
    my $source = 'EX=C'.$self->sourceCurrency;
    my $dateRange = 'Fr='.$fromSecondsSinceYearOne.'0000000&To='.$toSecondsSinceYearOne.'0000000';
    return $target.'&'.$source.'&'.'P=DateRange&'.$dateRange.'&CF=UnCompressed&CUF=Period&DS=Ascending&DT=NA'
}

sub serealizeCurrency{
    my $self = shift;

    my $targetCurrecy = $self->targetCurrencies;
    foreach(@$targetCurrecy){ $_ = "'$_'";} #Quote all strings
    return join(',', @$targetCurrecy);
}

=head1 AUTHOR

Guillermo Martinez, C<< <guillermo.marcial at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-currency-exchange-imf at rt.cpan.org>, or through
the web interface at L<https://rt.cpan.org/NoAuth/ReportBug.html?Queue=Currency-Exchange-IMF>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Currency::Exchange::IMF


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<https://rt.cpan.org/NoAuth/Bugs.html?Dist=Currency-Exchange-IMF>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Currency-Exchange-IMF>

=item * CPAN Ratings

L<https://cpanratings.perl.org/d/Currency-Exchange-IMF>

=item * Search CPAN

L<https://metacpan.org/release/Currency-Exchange-IMF>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

This software is Copyright (c) 2020 by Guillermo Martinez.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)


=cut

1; # End of Currency::Exchange::IMF
